nkms_eth/escrow.py

```diff
@@ -1,78 +1,132 @@
-    return escrow
-
-
-def get():
-    """
-    Returns an escrow object
-    """
-    return token.get(ESCROW_NAME)
-
-
-def sample(n: int = 10)-> List[str]:
-    """
-    Select n random staking Ursulas, according to their stake distribution
-    The returned addresses are shuffled, so one can request more than needed and
-    throw away those which do not respond
-    """
-    escrow = get()
-    n_select = int(n * 1.7)  # Select more ursulas
-    n_tokens = escrow.call().getAllLockedTokens()
-    duration = 10
-
-    for _ in range(5):  # number of tries
-        points = [0] + sorted(random.randrange(n_tokens) for _ in
-                              range(n_select))
-        deltas = [i - j for i, j in zip(points[1:], points[:-1])]
-        addrs = set()
-        addr = NULL_ADDR
-        shift = 0
-
-        for delta in deltas:
-            addr, shift = escrow.call().findCumSum(addr, delta + shift, duration)
-            addrs.add(addr)
-
-        if len(addrs) >= n:
-            addrs = random.sample(addrs, n)
-            return addrs
-
-    raise Exception('Not enough Ursulas')
+
+from populus.contracts.contract import PopulusContract
+
+from nkms_eth.token import NuCypherKMSToken
+from .blockchain import Blockchain
+
+addr = str
+
+
+class Escrow:
+    _contract_name = 'MinersEscrow'
+    hours_per_period = 1       # 24
+    min_release_periods = 1    # 30
+    max_awarded_periods = 365
+    min_allowed_locked = 10 ** 6
+    max_allowed_locked = 10 ** 7 * NuCypherKMSToken.M
+    reward = NuCypherKMSToken.saturation - NuCypherKMSToken.premine
+    null_addr = '0x' + '0' * 40
+
+    mining_coeff = [
+        hours_per_period,
+        2 * 10 ** 7,
+        max_awarded_periods,
+        max_awarded_periods,
+        min_release_periods,
+        min_allowed_locked,
+        max_allowed_locked
+    ]
+
+    class ContractDeploymentError(Exception):
+        pass
+
+    class NotEnoughUrsulas(Exception):
+        pass
+
+    def __init__(self, blockchain: Blockchain, token: NuCypherKMSToken, contract: PopulusContract=None):
+        self.blockchain = blockchain
+        self.contract = contract
+        self.token = token
+        self.armed = False
+        self.miners = list()
+
+    def __call__(self):
+        """Gateway to contract function calls"""
+        return self.contract.call()
+
+    def __eq__(self, other):
+        return self.contract.address == other.contract.address
+
+    def arm(self):
+        self.armed = True
+        return self
+
+    def deploy(self):
+        if not self.armed:
+            raise self.ContractDeploymentError('use .arm() to arm the contract, then .deploy().')
+        if self.contract:
+            class_name = self.__class__.__name__
+            message = '{} contract already deployed, use .get() to retrieve it.'.format(class_name)
+            raise self.ContractDeploymentError(message)
+
+        contract, deploy_txhash = self.blockchain.chain.provider.deploy_contract(self._contract_name,
```

Comment:
Dense.  And isn't this a very significant moment?

Maybe break it up into two lines, with names fitting the momentousness of executing this.